src/llama.cpp

```diff
@@ -19559,6 +19559,21 @@ static int32_t llama_chat_apply_template_internal(
         if (add_ass) {
             ss << "ASSISTANT:";
         }
+    } else if (tmpl == "alpaca" || (tmpl.find("### Instruction:") != std::string::npos && tmpl.find("<|EOT|>") == std::string::npos)) {
+        // meta-math/MetaMath-7B-V1.0
+        for (auto message : chat) {
+            std::string role(message->role);
+            if (role == "system") {
+                ss << message->content << "\n\n";
```

Comment:
In any cases, having `\n\n` after system message makes more sense to me. Could we delete the template for `deepseek` below and move it here?

```
} else if (tmpl == "alpaca" || tmpl == "deepseek" || (tmpl.find("### Instruction:") != std::string::npos && tmpl.find("<|EOT|>") == std::string::npos)) {
```

Reply:
If you look in the previous PR that I closed accidentally, then there are some examples of `phind-codellama` getting very confused when I tried to use the `EOS` token.

I am unsure how people use alpaca for creative writing, but am going to try it over the weekend using [mikupad](https://github.com/lmg-anon/mikupad), which seems to have forgotten to add any `EOS` tokens to any of their prompts (but nobody seems to have noticed or cared, and the models seem to have been writing fiction fine!?).

Reply:
Probably we can also patch `llama_chat_apply_template_internal` to take the EOS as an input. Since this function is used internally, we can change the signature whenever we want.

Do you think that will be a better solution?

> I am unsure how people use alpaca for creative writing, but am going to try it over the weekend using [mikupad](https://github.com/lmg-anon/mikupad), which seems to have forgotten to add any EOS tokens to any of their prompts (but nobody seems to have noticed or cared, and the models seem to have been writing fiction fine!?).

That sounds to me like a bit of mess when the world moving from single-turn to multi-turn template 1 year ago. Back then, no one care about putting EOS after model's response because we can match stop sequence (`### User:`). In fact, for this reason, I was skeptical to add alpaca template in the first place, since some models are trained solely with single-turn dataset.

Reply:
> Probably we can also patch `llama_chat_apply_template_internal` to take the EOS as an input (can be read from `llama_model`). Since this function is used internally, we can change the signature whenever we want.
> 
> Do you think that will be a better solution?

It may be worth looking at the model files of the few genuine multiturn Alpaca models to see if they even have a EOS token defined. Sadly it was the now deleted "wizard" models that mainly used this but some of the early "meta" models and `phind-codellama` also used it. 

> > I am unsure how people use alpaca for creative writing, but am going to try it over the weekend using [mikupad](https://github.com/lmg-anon/mikupad), which seems to have forgotten to add any EOS tokens to any of their prompts (but nobody seems to have noticed or cared, and the models seem to have been writing fiction fine!?).
> 
> That sounds to me like a bit of mess when the world moving from single-turn to multi-turn template 1 year ago. Back then, no one care about putting EOS after model's response because we can match stop sequence (`### User:`). In fact, for this reason, I was skeptical to add alpaca template in the first place, since some models are trained solely with single-turn dataset.

The mikupad Dec's just fixed all their templates after I pointed this out.